Initial design of Morello compartments library

0152la opened the pull request:
The functionality is rather limited at the moment, but in the interest of not having too big PRs to review, I think it's a good step to have a sync. Currently, there are two main files of interest:
* `src/switcher.S`, more or less copied from our previous `secure` [1] example;
* `src/manager.S`, containing new functions for, currently, initialization and adding compartments.
The current system follows the idea that we'll have one function of interest per compartment (although that can easily be changed). In `init_compartments`, we allocate memory for the switcher, including memory for hosting compartment capabilities which will be used for switching, as well as derive capabilities (DDC/PCC) for the switcher. In `add_compartment`, we allocate memory for a new compartment, as well as derive appropriate data (DDC) and executable (PCC) capabilities for the compartment and the executable entry point, then store these capabilities within the switcher's memory region (containing `comps_addr`).

In addition, there are two tests, `simple_init` and `simple_add`, which call the above functions, and perform some sanity checks over their functionality.

This will remain a draft until I get bors working with this repo.

[1] https://github.com/capablevms/cheri-examples/tree/master/hybrid/compartment_examples/inter_comp_call/secure

Commit:
Add basic infrastructure, and initial work

* usual `src`, `include`, `tests` setup
* use in-built cmake option for testing
* sketch out `init_compartments` and `add_compartment` functions

Commit:
Update `init_compartments` and `simple_init`

As discussed, we move towards a more abstract model, including memory
management in the assembly layer. This updates the initialiation
function to not require any user input, and setup memory appropriately,
for future use.

Add checks to `simple_init` to sanity check functionality.

Commit:
Finalize `simple_add` test

ltratt commented:
I might have misunderstood this but it *looks* like there are two global arrays `switcher_caps` and `comps_addr` which contain capabilities which any compartment can access at any time?

jacobbramley commented:
Were they intended to be sealed, perhaps?

0152la commented:
They will be lying outside the respective DDC of the compartments, so won't be accessible, but yes, I believe these could be sealed. Once I model actual switching here, I will be able to evaluate what is needed precisely. Currently, everything happening here would be in privileged/executive mode.

ltratt commented:
I am going to assert (which means: I'll be happy if you prove me wrong!) that the current API will always leak capabilities via the arrays. In https://github.com/capablevms/cheri-examples/issues/45 I tried doing something different which is that `compartment_create` gives you back a (probably sealed?) capability which uniquely identifies a compartment: it's then up to you who you give access to that capability to. I suspect that's the only way to make this API secure(able).

0152la commented:
Well, those arrays reside in some memory location, let's say 0x1000 to 0x2000. Suppose we construct compartments with memory allocated to them somewhere in regions 0x4000 to 0x40000. Therefore, any derived DDC for those compartments will be constrained within that region, meaning that when we will be executing code within some compartment, we know its DDC will not cover memory region 0x1000 to 0x2000. Do you disagree?

There is a secondary, orthogonal thing I thought about, which is having all executable entry-points within their respective compartments, and have a stub entry point for each compartment, which is called by the switcher, but that's something to be done in the future, I think. That would be closer to what you're proposing in that issue.

ltratt commented:
> Well, those arrays reside in some memory location, let's say 0x1000 to 0x2000. Suppose we construct compartments with memory allocated to them somewhere in regions 0x4000 to 0x40000. Therefore, any derived DDC for those compartments will be constrained within that region, meaning that when we will be executing code within some compartment, we know its DDC will not cover memory region 0x1000 to 0x2000. Do you disagree?

It's not that I disagree per se, but that means that `simple_add` would presumably have to be rewritten to use a quite different API (e.g. it accesses array elements directly at https://github.com/capablevms/cheri-morello-compartmentalisation/pull/1/files#diff-381d671f972d99b39707fab1f10cb11be8ea79e414a0f5790ea3991be7472ef3)?

0152la commented:
Well, that would be a different example, if we want to access elements from within a compartment. The two tests do everything from a fully privileged status, as presumably we expect the functionality of creating/deleting compartments to be called from within such a context. To emphasise, the two current tests do not showcase any execution from within compartments, so there are no security guarantees to be checked within these, I don't think.

Commit:
Add bors + buildbot support, and legal stuff

ltratt commented:
I'm going to push you to reconsider: I *think* the array approach is fundamentally broken because either:

1. We directly leak every compartments capability (as at the moment).
2. We indirectly leak it by allowing people to say something like "get_capability(1);"

I'm going to go further and (in full awareness of how arrogant this is on my part!) assert that https://github.com/capablevms/cheri-examples/issues/45 has an API which is secure now and will remain so in the future.

Warning: I might be wrong!

0152la commented:
I disagree, but what I suggest we do is that we wait until I add a test which spawns more compartments, and actually does inter-compartment calls (which I'm unsure precisely how it'll look like at the moment) and then we can model attempts at leaking these stored capabilities. Is that agreeable?

ltratt commented:
Before you do that, I think you need to prove that compartment *A* can never get access to compartment *B*'s capability (unless someone explicitly gives *A* such a capability). If you can do that, then I'm willing to let go.

ltratt commented:
After an offline chat, I had misunderstood that the examples are (in essence) running "inside" the (privileged) compartment switcher. My comments above thus don't make fully sense. Mea culpa!

0152la commented:
bors try

0152la commented:
bors try-

0152la commented:
bors ping

0152la commented:
bors try

0152la commented:
bors try

0152la commented:
bors try-

0152la commented:
bors try

0152la commented:
bors try

0152la commented:
bors try

0152la commented:
bors try

0152la commented:
bors try

0152la commented:
bors try

0152la commented:
bors try

0152la commented:
bors try

0152la commented:
bors try

0152la commented:
bors try

0152la commented:
bors try

0152la commented:
bors try

0152la commented:
bors try

0152la commented:
bors try

0152la commented:
bors try

0152la commented:
bors try

0152la commented:
bors try

0152la commented:
Ready for review @ltratt (not pinging Jacob as he's on holiday).

I will be fixing the solution I've chosen, as I can't be bothered to find a "proper" way of doing this. I have other solutions, but they would involve buildbot changes, and this works for now. They can be discussed at a future time, if appropriate.

ltratt commented:
Can you squash the new commits as you see appropriate (please don't squash any commit including/before https://github.com/capablevms/cheri-morello-compartmentalisation/pull/1/commits/96a72ee6eec2df7bf431e48c6b3fec445960dfd8) and force push to make it easier to review? Thanks!

0152la commented:
Squashed the boilerplate commits away.

ltratt commented on COPYRIGHT:
```diff
@@ -0,0 +1,17 @@
+The following contributors wish to explicitly make it known that the copyright
+of their contributions is retained by an organisation:
+
+    Jacob Bramley <jacob.bramley@arm.com>:
```
I think (offline) Jacob was right to say that people committing from an institutional address don't need to be listed here. It's just awkward people like me!

jacobbramley replied:
It's true, such contributors do not need to be explicitly listed, but it's also harmless to do so. I have no objection either way. When I made my earlier (offline) comments, I was confused about the nature of this file, otherwise I wouldn't have said anything.

ltratt commented:
> I will be fixing the solution I've chosen, as I can't be bothered to find a "proper" way of doing this.

I'm not sure I understand this. Can you expand please?

0152la commented:
To give a bit of background, before I started integrating bors in this project, I already was having a working-state testing infrastructure. Ideally, I'd be able to just apply that existing infrastructure with no changes, in the testing environment (or update it in case things are improperly engineered). However, it seems I need to add these `StrictHostKeyChecking no` options to commands in my test script [1], just to get it working on our buildbot worker container. I dislike this solution of having to modify the generic testing infrastructure, but after having spent a few hours getting annoyed at buildbot, I just decided to "make it work", so we can make progress with the PR.

[1] https://github.com/capablevms/cheri-morello-compartmentalisation/pull/1/files#diff-a846d9efb204234acca8da8ca8a4ad031932989e1da0ef08960fd78ec3d60fa2R16-R17

ltratt commented:
I'm confused: are you saying the testing infrastructure for (say) cheri-examples doesn't work here or ... ?

0152la commented:
There are three components to using our buildbot infrastructure:
* bors/buildbot communication
* spawning a QEMU CHERI instance
* running tests for the project

The first two, I've reused from `cheri-examples`. The last, of course, should be project specific, and by testing infrastructure above, I meant the testing of this project, by its own. In order to move that in our buildbot setup, I had to modify it slightly to make it compatible, when I think those requested changes should be done on the buildbot side (or understand if there is a better solution, since I don't fully understand the core cause), but that is another conversation.

ltratt commented:
Ah I see! My guess is that the "running tests for the project" thing is necessarily repo specific (even if it's often very similar from repo to repo). Might be worth discussing with @drpdr in case I'm wrong though!

Please squash the PR overall as you see fit.

0152la commented:
I've been discussing with Pino, and proposed something, but we need to think it through first.

By squashing, you mean this is still too many commits? I've squashed everything after https://github.com/capablevms/cheri-morello-compartmentalisation/commit/96a72ee6eec2df7bf431e48c6b3fec445960dfd8, as per your above comment. I can squash the ones before, as well, to just have a monolithic commit with all the implementation stuff.

ltratt commented:
You should squash as you see fit, but yes please squash things back to the beginning of the PR so that I can merge this. That doesn't mean you have to squash all the commits away (but e.g. I doubt 20 commits in history is useful for what this PR does! What the right number is, though, I leave up to you).

0152la commented:
I think this should be a bit cleaner, while maintaining a high-level overview.

ltratt commented on src/manager.S:
```diff
@@ -0,0 +1,169 @@
+    add       x4, x4, #1
+    str       x4, [x3]
+
+    // Update switcher DDC
```
Is there a reason this is commented?

0152la replied:
Ah right. Initially I was thinking if the switcher DDC should be dynamic, meaning that as we create compartments, therefore associated capabilities, we lengthen the visible space for the switcher (and vice-versa for deleting). But then I thought that we allocate some maximum memory at the start anyway (and currently there is no handling for what happens if we need more memory than that max), so wouldn't dynamically changing the DDC just add complexity, without much gain?

ltratt replied:
Good point. In the longer term, we will need to allow the switcher to create (and delete!) an arbitrary number of compartments. However, that shouldn't change the external API, so it's fine if we only allow a fixed number at first (but if we do that, we should add a check so that we abort the process if the user tries to exceed that fixed number).

0152la replied:
Implemented in https://github.com/capablevms/cheri-morello-compartmentalisation/pull/1/commits/61b9d8aedebd12dc9032854df65191f54ca1cc32.

ltratt replied:
Looks good to me -- but can we also delete the commented out code if it's not needed?

0152la replied:
Done in https://github.com/capablevms/cheri-morello-compartmentalisation/pull/1/commits/7e163169774b6b5844459b33b570cc87c40d3cc4 for now. I guess we can keep the idea in mind, if we do want to take that approach.

ltratt replied:
Lots of new code seems to have crept in with https://github.com/capablevms/cheri-morello-compartmentalisation/commit/7e163169774b6b5844459b33b570cc87c40d3cc4 ?

0152la replied:
Fixed in https://github.com/capablevms/cheri-morello-compartmentalisation/pull/1/commits/3defc97e5183f622526b1ebc7b98882670f1c985. (that was painful)

Commit:
Polish pass over `init` and `add` implementations

* Add checks to ensure we don't add more compartments than we allocated
  space for
* Small optimisations and unneeded code removal

Remove comment

Simplify

Replace multiple instruction with a single one

Remove unimplemented stub

Optimization

Better register usage

Fix

Fix

Forgot to update checked register

Removed unused function

Remove an unneeded call

Remove unneeded instruction

ltratt commented:
Please squash.

0152la commented:
I suggest we first wait for @jacobbramley to give us some feedback, once he's back from his holiday. I'll keep working on my fork, so this not being merged won't be a blocker.

jacobbramley commented on .gitignore:
```diff
@@ -0,0 +1,6 @@
+build/
+
+**/*.swp
```
Is that for Vim swap files? Isn't it better to ignore those in a user `~/.gitignore` (or similar)?

0152la replied:
I like having robust per-project configs, so when I work with them remotely, the settings are maintained. Are you suggesting we remove the swap files here by any chance?

jacobbramley replied:
I was, but it's pretty clearly bikeshedding so please disregard :-)

jacobbramley commented on CMakeLists.txt:
```diff
@@ -0,0 +1,24 @@
+cmake_minimum_required(VERSION 3.16)
```
We dropped cmake for `cheri-examples`. I don't remember all the reasoning (other than that it just wasn't required). Is it necessary here? I have no particular opinion one way or the other, except that simple is good.

0152la replied:
Personally, I understand `cmake` better than `make`, so I prefer it. If there is desire to have `make` consistently, I have no strong feelings either way, although do prefer `cmake` for myself.

jacobbramley commented on src/manager.S:
```diff
@@ -0,0 +1,165 @@
+asm_call_wrapper:
+    // Save `x0` so we can temporarily use it
+    cvtp      c0, x0
+    str       c0, [sp]
```
This will overwrite the caller's stack. Storing to `[sp]` is a _poke_, not a _push_. The processor won't complain — AAPCS64 is just a software convention and we're already deviating from it — but I'm fairly sure that it's not what you want.

```
    str c0, [sp, #-16]!  // Push-like behaviour.
    ldr c0, [sp], #16    // Pop-like behaviour.
```

Corresponding sequences elsewhere in the patch use the instructions that I would expect, so this appears to be an isolated bug (or misunderstanding on my part).

On AArch64 we often coalesce multiple pushes or pops to avoid repeatedly updating `sp`. I wrote about this in [a blog](https://community.arm.com/arm-community-blogs/b/architectures-and-processors-blog/posts/using-the-stack-in-aarch32-and-aarch64); nothing fundamentally changes with Morello, except that you can push individual (16-byte) C registers, which simplifies things sometimes.

Also, see my next comment :-)

0152la replied:
Right, I've done this in my other functions, but haven't managed to fix this up.

jacobbramley commented on src/manager.S:
```diff
@@ -0,0 +1,165 @@
+    cvtp      c0, x0
+    str       c0, [sp]
+
+    // Derive `clr` (in case asm function does something weird with `PCC`)
```
Ok, that's a very useful comment, because my first thought otherwise is "why are you preserving PCC metadata and not just `lr`?" Thanks!

Some observations:
* You don't need to do this for both `c0` and `lr`. It's not harmful, but might be clearer if you do it only for `lr` (since you can only restore `PCC` from one anyway), and then there's an obvious free stack slot for something else, should you need it later.
* I'm 99% sure that you don't want to use `swp`, which does an atomic swap. In this case, it will swap `c0` (containing the `lr` with the `PCC`'s metadata) with `[sp]` (containing the target function pointer with the `PCC`'s metadata). However, assuming no other threads are fiddling with the same memory, you had the target function pointer in a register a few instructions ago anyway.
* `cvtp` explicitly derives a capability from PCC. If you need to save or handle a function pointer, that's what you want. However, if you're only going to branch to it (like `blr c0`) then you might as well just branch to the X register (`blr x0`).

Taking those points into account, you probably want something this shape:

```
asm_call_wrapper:
    cvtp  clr, lr
    str   clr, [sp, #-32]!
    str   x0, [sp, #16]
    // Stack layout here:  [       unused      ]
    //                     [   x0 (target fn)  ]
    //                     [    clr (high64)   ]
    //               sp -> [    clr (low64)    ]

    blr    x0  // New PCC is implicitly derived from current PCC, using the address in x0.
    ldr    clr, [sp], #32
    ret    clr
```

That's obviously untested. Use at your own risk, etc.

0152la replied:
So the reasoning behind this function was to make my life easier, not needing to save some registers as I'm coming from C-world into asm, and just use this to call asm functions from C. I think it made it harder, as not only would the call point look bad (couldn't call an asm function directly, needed to go through this function), I think setting the registers correctly to the inner called functions is just a waste of resources. So I'll just get rid of it for now [1]. The asm functions are already able to be called directly.

The note about `cvtp` and `blr` is useful. It makes sense, I just hadn't thought of it before. Something to keep in mind.

[1] https://github.com/capablevms/cheri-morello-compartmentalisation/pull/1/commits/13a35e5ef17a41b99e0ffd4e0e6734450df0da76

jacobbramley commented on src/manager.S:
```diff
@@ -0,0 +1,165 @@
+.global del_compartment
+
+/**
+ * Wrapper to store/restore state when coming from C
```
What this _really_ does is call `x0`, with the current `PCC`. What sets it apart from a trivial trampoline is that it attempts to guarantee that the `PCC` on return is the same as it was on entry. It doesn't save/restore anything else.

Note, however, that this isn't a _security_ guarantee: the callee (at `x0`) could break CFI by returning to the `ret clr`, or directly to the caller of this function (using some other capability), and change the `PCC` that way.

0152la replied:
I'm having a hard time seeing on the github interface what you are referring to, so I'll assume it's `asm_call_wrapper`. This function (and all functions in `manager.S`) are meant to only be called from a privileged environment, and I initially designed it to not have to save the `clr` in all internal functions, and to setup the registers properly for the eventual function call (in `x0`) it was supposed to call. I abandoned developing this, and, as the number of privileged functions are limited, it might be easier to just drop this? At the moment it is unused, so maybe just for that it should be deleted from this PR.

jacobbramley replied:
If it's not used, then it's probably best to drop it, if only to simplify the analysis. (I hadn't spotted that it's not used.)

0152la replied:
The function `asm_call_wrapper` is now deleted in the latest version. Guess it's kinda hard to see with this Github UI.

jacobbramley commented on src/manager.S:
```diff
@@ -0,0 +1,165 @@
+    mov       x0, #COMP_SIZE
+    mov       x1, #MAX_COMP_COUNT
+    mov       x2, #32    // size of the 2 switcher capabilities
+    madd      x1, x0, x1, x2
```
I think that's correct, but you can just write a static expression and do it in a single instruction:
```
mov  x1, #(32 + (COMPSIZE * MAX_COMP_COUNT))
```

It'll fail if it's too big for a single `mov`, though. In that case, you can use a [special immediate syntax](https://sourceware.org/binutils/docs-2.36/as.html#AArch64_002dRelocations):

```
#define ALLOC_SIZE  (32 + (COMPSIZE * MAX_COMP_COUNT))
movz x1, #:abs_g1:ALLOC_SIZE
movk x1, #:abs_g0_nc:ALLOC_SIZE
```

0152la replied:
I didn't know that; thanks! Updated in https://github.com/capablevms/cheri-morello-compartmentalisation/pull/1/commits/eec1ee0f43abae64bbba5ba35ee7ea93a4747b49.

jacobbramley commented on src/manager.S:
```diff
@@ -0,0 +1,165 @@
+
+    // Store (DDC, PCC) at `switcher_caps`
+    ldr       x1, switcher_caps
+    stp       c2, c3, [x1], #32
```
This is odd. From the comment, I'd expect `adr x1, switcher_caps`, but actually I think this needs another level of indirection. Perhaps `switcher_caps[0,1]` would be clearer.

Note that the value loaded here into `x1` should already be in `x0` (from the `mmap` call), so this load can be elided.

The `, #32` is clever and compact!

0152la replied:
So `adr x1, switcher_caps` would get the address where the `switcher_caps` variable is stored. What I want is the value stored at that address. I think the comment is slightly unclear. You are correct that we have it in `x0`. I've updated this in https://github.com/capablevms/cheri-morello-compartmentalisation/pull/1/commits/67ff6da6fe76bebab07d93ff26732ad1e188ee8d.

0152la replied:
I see what I did there; I was modifying the value of `x1`, but still needed the return address to be in `x0` unmodified, as I want to return that. But still, can replace the load with a move, which I assume is better. Fixed in https://github.com/capablevms/cheri-morello-compartmentalisation/pull/1/commits/41cf6609258f7987272a06c0577a547cf52d5174.

jacobbramley replied:
Yes, sorry, I was trying to point out that the comment was unclear. I think the code was correct (and yes, a `mov` is almost certainly cheaper than an `ldr`).

jacobbramley commented on src/manager.S:
```diff
@@ -0,0 +1,165 @@
+
+    // Increment counter
+    adr       x3, comps_cnt
+    ldr       x4, [x3]
```
You already had this value in `x3` (though the `adr` overwrote it). Perhaps `adr` into something else, then you can elide this load.

0152la replied:
Good spot. Updated in https://github.com/capablevms/cheri-morello-compartmentalisation/pull/1/commits/239f9c5f382bef90fec77e884f2ed58556fa85eb.

jacobbramley commented on src/manager.S:
```diff
@@ -0,0 +1,165 @@
+ * @param c0 ID of compartment to be deleted
+ */
+.type del_compartment, "function"
+del_compartment:
```
Unimplemented? Should this at least return so it doesn't break control flow?

0152la replied:
Removed in https://github.com/capablevms/cheri-morello-compartmentalisation/pull/1/commits/26b1f3e78cbf2b29cfd52c95f42eb990a99d6fc4.

jacobbramley commented on src/switcher.S:
```diff
@@ -0,0 +1,134 @@
+    mov      x0, #0
+    cvtp     clr, lr
+    b        switch_compartment
+    ret      clr
```
Do we ever come back here? The `b` looks like a tail call, so is this instruction necessary?

0152la replied:
Ah, good point. We can just drop through. I copy pasted this code from the previous switcher iteration (and honestly shouldn't have included it in this PR), but for completeness, I updated it in https://github.com/capablevms/cheri-morello-compartmentalisation/pull/1/commits/08f834ff048f627173e5959fa7fe7d54a2107c66.

jacobbramley commented on src/switcher.S:
```diff
@@ -0,0 +1,134 @@
+
+/** Code to perform actual switch
+ *
+ * @param x0 ID (in `comps` array) of compartment to switch to
```
It also expects the switcher `DDC` in `c29`, and it assumes that its _address_ is set to some data area too. (The `DDC` address is usually ignored.)

0152la replied:
Yes, I've documented this in the heavily modified local copy I'm working on right now. Thanks for spotting that.

jacobbramley commented on src/switcher.S:
```diff
@@ -0,0 +1,134 @@
+ */
+.type switcher_entry, "function"
+switcher_entry:
+    mov      c29, c0
```
Note that `c29` is callee-saved, so this should preserve it (since this is called from C code).

A simple fix for now would be to pass it in some caller-saved register not used for arguments (e.g. c10). However, I think we picked `c29` earlier because it's also special for `ldrblr` etc. Currently, we aren't using that, but maybe there's a development plan for it.

0152la replied:
That's a good point, having to save `c29`. Let's leave this conversation for the PR where the switcher is actually used. I'm sure there's more correctness issues to be discussed then.

jacobbramley replied:
I've no objection in principle, but I worry that it'll be forgotten and bite us later. Bugs like these tend to go unnoticed because simple calling code might continue to work anyway.

The safer approach would be `s/c29/c10/` now (here and in `switch_compartment`). If we want to use `ldpblr` later, we can revisit this.

0152la replied:
I can do this for completion here, but the problem is I'd be needing to rebase those changes into my local branches, which are complete (and include `ldpblr` use appropriately). If you think it's worth it for having this PR more "complete", then I can do that, but this code is due to be overwritten shortly anyway. And I know I cut some corners regarding saving registers in the implementation I have locally, so that's something I definitely want to fix over in that PR; I definitely won't let it go.

jacobbramley replied:
I'm not insisting on anything, so handle this as you prefer :-)

jacobbramley commented on src/switcher.S:
```diff
@@ -0,0 +1,134 @@
+    // Install compartment DDC
+    msr       DDC, c1
+
+    // Save old DDC (c2), old SP (x12), old CLR (clr) on stack
+    stp       c2, clr, [sp, #-48]!
+    str       x12, [sp, #32]
```
Doesn't this make them reachable from the callee compartment? If so, why isn't it a security hole?

0152la replied:
Aha, we found a security hole in what we considered a secure implementation of switching [1]. Indeed, these should be saved before we swap the DDC. I will keep this in mind as I work on the switcher. We could probably add a test which shows this can be done on the prior example, then fix the security hole.

[1] https://github.com/capablevms/cheri-examples/blob/master/hybrid/compartment_examples/inter_comp_call/secure/switch_compartment.s#L35-L40

jacobbramley commented on src/switcher.S:
```diff
@@ -0,0 +1,134 @@
+    blr       c0
+
+    // Clean capabilities left in the return value.
+    mov       w0, w0
```
This is redundant with the next line (which calls `clean` and thereby clears `w0` anyway).

0152la replied:
Fixed in https://github.com/capablevms/cheri-morello-compartmentalisation/pull/1/commits/80e78ca3bbce91bc3f4bac230f4ecbded8d18bde.

0152la commented:
I think the discussion around this PR has reached a fixed point, and we can go ahead and merge it.

ltratt commented:
Please squash.

0152la commented:
Squashed.

ltratt commented:
bors r+
